osquery/tables/system/linux/md_tables.cpp

```diff
@@ -0,0 +1,733 @@
+
+    if (std::stoi(devMajor) == major && std::stoi(devMinor) == minor) {
+      devName = udev_device_get_property_value(device, "DEVNAME");
+      return true;
+    }
+
+    return false;
+  });
+
+  return devName;
+}
+
+std::string MD::getSuperblkVersion(std::string arrayName) {
+  std::string version;
+
+  walkUdevDevices("block", [&](udev_device* const& device) {
+    const char* devName = udev_device_get_property_value(device, "DEVNAME");
+
+    if (strcmp(arrayName.c_str(),
+               &devName[strlen(devName) - arrayName.length()]) == 0) {
+      version = udev_device_get_property_value(device, "MD_METADATA");
+      return true;
+    }
+
+    return false;
+  });
+
+  return version;
+}
+
+/**
+ * @brief resolves MD disk state field to string representation
+ *
+ * @param  state state field of mdu_disk_info_t
+ *
+ * @return stringified state
+ */
+std::string getDiskStateStr(int state) {
+  // If state is 0, which is undefined, we assume recoverying, as this is all
+  // have seen in the wild
+  if (state == 0)
+    return "recovering";
+
+  std::string s;
+
+  if ((1 << MD_DISK_FAULTY) & state)
+    s += "faulty ";
+
+  if ((1 << MD_DISK_ACTIVE) & state)
+    s += "active ";
+
+  if ((1 << MD_DISK_SYNC) & state)
+    s += "sync ";
+
+  if ((1 << MD_DISK_REMOVED) & state)
+    s += "removed ";
+
+  if ((1 << MD_DISK_WRITEMOSTLY) & state)
+    s += "writemostly ";
+
+#ifdef MD_DISK_FAILFAST
+  if ((1 << MD_DISK_FAILFAST) & state)
+    s += "failfast ";
+#endif
+
+#ifdef MD_DISK_JOURNAL
+  if ((1 << MD_DISK_JOURNAL) & state)
+    s += "journal ";
+#endif
+
+#ifdef MD_DISK_CANDIDATE
+  if ((1 << MD_DISK_CANDIDATE) & state)
+    s += "spare ";
+#endif
+
+#ifdef MD_DISK_CLUSTER_ADD
+  if ((1 << MD_DISK_CLUSTER_ADD) & 1)
+    s += "clusteradd ";
+#endif
+
+  boost::algorithm::trim(s);
+  return s;
+}
+
+/**
+ * @brief resolves superblock state field to string representation
+ *
+ * @param  state state field of mdu_array_info_t
+ *
+ * @return stringified state
+ */
+std::string getSuperBlkStateStr(int state) {
+  if (state == 0)
+    return "unknown";
+
+  std::string s;
+
+#ifdef MD_SB_CLEAN
+  if ((1 << MD_SB_CLEAN) & state)
+    s += "clean ";
+#endif
+
+#ifdef MD_SB_ERRORS
+  if ((1 << MD_SB_ERRORS) & state)
+    s += "errors ";
+#endif
+
+#ifdef MD_SB_BBM_ERRORS
+  if ((1 << MD_SB_BBM_ERRORS) & state)
+    s += "bbm_errors ";
+#endif
+
+#ifdef MD_SB_BLOCK_CONTAINER_RESHAPE
+  if ((1 << MD_SB_BLOCK_CONTAINER_RESHAPE) & 1)
+    s += "container_reshape ";
+#endif
+
+#ifdef MD_SB_BLOCK_VOLUME
+  if ((1 << MD_SB_BLOCK_VOLUME) & 1)
+    s += "block_activation ";
+#endif
+
+#ifdef MD_SB_CLUSTERED
+  if ((1 << MD_SB_CLUSTERED) & 1)
+    s += "clustered ";
+#endif
+
+#ifdef MD_SB_BITMAP_PRESENT
+  if ((1 << MD_SB_BITMAP_PRESENT) & 1)
+    s += "bitmap_present ";
+#endif
+
+  boost::algorithm::trim(s);
+  return s;
+}
+
+// For use with unique_ptr of file close as a hacky way of preventing fd leaks
+auto fClose = [](int* fd) { close(*fd); };
```

Comment:
Can this be moved inside of the functions it's being used in? If not can you find a way to make it a `const` with the `k` prefix notation?